File: Python/524. Longest Word in Dictionary through Deleting.py
class Solution:
    def findLongestWord(self, s: str, d: List[str]) -> str:
        def isSubsequence(x, y):
            i = j = 0
            while i < len(x) and j < len(y):
                if x[i] == y[j]:
                    i += 1
                j += 1
            return i == len(x)
        max_str = ""
        for ds in d:
            if isSubsequence(ds, s):
                if len(ds) > len(max_str) or len(ds) == len(max_str) and ds < max_str:
                    max_str = ds
        return max_str
            

File: 2021/524. Longest Word in Dictionary through Deleting.cpp
class Solution {
public:
    string findLongestWord(string s, vector<string>& d) {
        vector<int> indexes(d.size(), 0);
        for (int i = 0; i < s.length(); i++) {
            for (int j = 0; j < d.size(); j++) {
                if (indexes[j] < d[j].length() && s[i] == d[j][indexes[j]]) {
                    indexes[j]++;
                }
            }
        }
        
        string longestWord;
        for (int i = 0; i < d.size(); i++) {
            if (indexes[i] == d[i].length() && (d[i].length() > longestWord.length() || (d[i].length() == longestWord.length() && d[i] < longestWord))) {
                longestWord = d[i];
            }
        }            
        return longestWord;
    }
};

class Solution {
public:
    string findLongestWord(string s, vector<string>& d) {        
        int maxIndex = -1;
        for (int di = 0; di < d.size(); di++) {
            string sub = d[di];
            int i = 0;
            for (int j = 0; j < s.length(); j++) {
                if (s[j] == sub[i]) i++;
            }            
            if (i >= sub.length() && (maxIndex == -1 || (sub.length() > d[maxIndex].length() || (d[maxIndex].length() == sub.length() && sub < d[maxIndex])))) {
                maxIndex = di;
            }
        }        
        
        return maxIndex == -1 ? "" : d[maxIndex];
    }
};